packages/gatsby/src/recipes/graphql.js

```diff
@@ -0,0 +1,284 @@
+  await asyncForEach(plan, applyStep(plan))
+  setTimeout(() => {
+    emitOperation(`success`, plan)
+  }, 1)
+}
+
+const OperationStateEnumType = new GraphQLEnumType({
+  name: `OperationStateEnum`,
+  values: {
+    RUNNING: { value: `progress` },
+    SUCCESS: { value: `success` },
+    ERROR: { value: `error` },
+  },
+})
+
+const OperationType = new GraphQLObjectType({
+  name: `Operation`,
+  fields: {
+    state: { type: OperationStateEnumType },
+    data: { type: GraphQLString },
+  },
+})
+
+const NPMPackageScriptType = new GraphQLObjectType({
+  name: `NPMPackageScript`,
+  fields: {
+    name: { type: GraphQLString },
+    script: { type: GraphQLString },
+  },
+})
+
+const NPMPackageType = new GraphQLObjectType({
+  name: `NPMPackage`,
+  fields: {
+    name: { type: GraphQLString },
+    version: { type: GraphQLString },
+    path: { type: GraphQLString },
+    scripts: { type: new GraphQLList(NPMPackageScriptType) },
+  },
+})
+
+const GatsbyConfigPluginType = new GraphQLObjectType({
+  name: `GatsbyConfigPlugin`,
+  fields: {
+    name: { type: GraphQLString },
+    version: { type: GraphQLString },
+  },
+})
+
+const rootQueryType = new GraphQLObjectType({
+  name: `Root`,
+  fields: () => {
+    return {
+      npmPackage: {
+        type: NPMPackageType,
+        args: {
+          name: {
+            type: GraphQLString,
+          },
+        },
+        resolve: async (_, args) => {
+          // TODO: peer/dev
+          const { dependencies } = await readPackage()
+          const version = dependencies[args.name]
+          return { name: args.name, version }
+        },
+      },
+      allNpmPackage: {
+        type: new GraphQLList(NPMPackageType),
+        resolve: async () => {
+          const { dependencies } = await readPackage()
+
+          return Object.entries(dependencies).map(([name, version]) => {
+            return {
+              name,
+              version,
+            }
+          })
+        },
+      },
+      allNpmPackageScripts: {
+        type: new GraphQLList(NPMPackageScriptType),
+        resolve: async () => {
+          const { scripts } = await readPackage()
+          return Object.entries(scripts).map(([name, script]) => {
+            return {
+              name,
+              script,
+            }
+          })
+        },
+      },
+      allGatsbyConfigPlugin: {
+        type: new GraphQLList(GatsbyConfigPluginType),
+        resolve: async () => {
+          const plugins = gatsbyPluginResource.read({ root })
+          return plugins.map(async name => {
+            return { name }
+          })
+        },
+      },
+    }
+  },
+})
+
+const rootMutationType = new GraphQLObjectType({
+  name: `Mutation`,
+  fields: () => {
+    return {
+      createOperation: {
+        type: GraphQLString,
+        args: {
+          commands: { type: GraphQLString },
+        },
+        resolve: (_data, args) => {
+          applyPlan(JSON.parse(args.commands))
```

Comment:
This should be a proper input type, not JSON.

Reply:
Generally, if we are reading recipes from files, what's the point of passing the MDX content from the file, instead of passing eg the file path and then reading it here?